experimental/cache_analysis.py

```diff
@@ -0,0 +1,139 @@
+from matplotlib.pyplot import figure
+import matplotlib.pyplot as plt
+import io
+import numpy as np
+import pandas as pd
+from statistics import mean, median
+import collections
+import csv
+import seaborn as sns
+
+""" This functions draws two types of plots for cases where each x value has only one y value.
+
+Here it is used to draw plots for stats (ie mean and median of all runs).
+It can be used for drawing each run individually.
+"""
+def draw_single_run_plot(data, name):
+  figure(num=None, figsize=(10, 8), dpi=80, facecolor='w', edgecolor='k')
+
+  x = list(data.keys())
+  y = list(data.values())
+
+  plt.subplot(211)
+  plt.plot(x,y)
+  plt.subplot(212)
+  plt.scatter(x,y)
+  plt.savefig(name)
+  plt.cla()
+
+
+""" This functions draws boxplot for all data points.
+"""
+def draw_boxplot(data, x, name):
+  fig = plt.figure(figsize=(30, 10))
+  ax = fig.add_subplot(111)
+  boxes = []
+  for l in data:
+    c = 0
+    for i in l:
```

Comment:
This whole bit can be replaced as so:

```py
for i, time in enumerate(l):
  boxes.append([x[i], time])
```

(Trying to use a more descriptive name than `i` for the time taken, IIUC.)

`enumerate` is a really cool function -- it yields (index, item) pairs (where its input is just an iterable of `item`), so it lets you avoid manually incrementing counters like `c` here.

https://docs.python.org/3/library/functions#enumerate